notebooks/community/aisamples/AIsample - Uplift Modelling.ipynb

```diff
@@ -0,0 +1,882 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "### Model evaluation\n",
+    "\n",
+    "Since actial uplift cannot be observed for each individual, we measure the uplift over a group of customers.\n",
+    "\n",
+    "- **Uplift Curve**: plots the real cumulative uplift across the population"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "test_pred_pandas_df = test_pred_df.toPandas()\n",
+    "test_pred_pandas_df.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "source": [
+    "First, we define some helper functions to plot uplift curve"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "jupyter": {
+     "outputs_hidden": false,
+     "source_hidden": false
+    },
+    "nteract": {
+     "transient": {
+      "deleting": false
+     }
+    }
+   },
+   "outputs": [],
+   "source": [
+    "def uplift_rank(uplift_df, treatment_col, label_col, uplift_col):\n",
+    "    # Rank the data by the uplift score\n",
+    "    ranked = pd.DataFrame(\n",
+    "        {\"treatment\": [], \"label\": [], \"uplift_score\": [], \"ranked_uplift\": []}\n",
+    "    )\n",
+    "    ranked[\"treatment\"] = uplift_df[treatment_col]\n",
+    "    ranked[\"label\"] = uplift_df[label_col]\n",
+    "    ranked[\"uplift_score\"] = uplift_df[uplift_col]\n",
+    "    ranked[\"ranked_uplift\"] = ranked.uplift_score.rank(pct=True, ascending=False)\n",
+    "    ranked = ranked.sort_values(by=\"ranked_uplift\").reset_index(drop=True)\n",
+    "    return ranked\n",
+    "\n",
+    "\n",
+    "def uplift_eval(ranked):\n",
+    "    uplift_df = ranked.copy()\n",
+    "    # Using Treatment and Control Group to calculate the uplift (Incremental gain)\n",
+    "    C, T = sum(ranked.treatment == 0), sum(ranked.treatment != 0)\n",
+    "    ranked[\"cr\"] = ranked.label\n",
+    "    ranked[\"tr\"] = ranked.label\n",
+    "    ranked.loc[ranked.treatment != 0, \"cr\"] = 0\n",
+    "    ranked.loc[ranked.treatment == 0, \"tr\"] = 0\n",
+    "    ranked[\"cr/c\"] = ranked.cr.cumsum() / C\n",
+    "    ranked[\"tr/t\"] = ranked.tr.cumsum() / T\n",
+    "    # Calculate and put the uplift value into dataframe\n",
+    "    uplift_df[\"uplift\"] = round(ranked[\"tr/t\"] - ranked[\"cr/c\"], 5)\n",
+    "\n",
+    "    # Add q0\n",
+    "    q0 = pd.DataFrame({\"ranked_uplift\": 0, \"uplift\": 0, \"treatment\": None}, index=[0])\n",
+    "    uplift_df = pd.concat([q0, uplift_df]).reset_index(drop=True)\n",
+    "\n",
+    "    return uplift_df\n",
```

Comment:
nit: is there any way to stay in spark for this? I know the rank operation might be tough in spark but if you can do the other steps in spark and then .toPandas().rank() at the very end it might be cleaner. No worries if you try and cant find a cleaner way though